README.rst
Happy Eyeballs in Python (RFC 6555)
===================================

.. image:: https://github.com/sethmlarson/rfc6555/actions/workflows/ci.yml/badge.svg
    :target: https://github.com/sethmlarson/rfc6555/actions/workflows/ci.yml

.. image:: https://codecov.io/gh/sethmlarson/rfc6555/branch/master/graph/badge.svg?token=Xn7oQWErjc
    :target: https://codecov.io/gh/sethmlarson/rfc6555    

.. image:: https://img.shields.io/pypi/v/rfc6555.svg?style=flat-square
    :target: https://pypi.python.org/pypi/rfc6555

Synchronous Python implementation of the Happy Eyeballs Algorithm described in `RFC 6555 <https://tools.ietf.org/html/rfc6555>`_.
Provided with a single file and dead-simple API to allow easy vendoring
and integration into other projects.

Abstract
--------

When a server's IPv4 path and protocol are working, but the server's
IPv6 path and protocol are not working, a dual-stack client
application experiences significant connection delay compared to an
IPv4-only client.  This is undesirable because it causes the dual-
stack client to have a worse user experience.  This document
specifies requirements for algorithms that reduce this user-visible
delay and provides an algorithm.

Installation
------------

 .. code-block:: bash

    $ python -m pip install rfc6555

Usage
-----

The main API for the ``rfc6555`` module is via ``rfc6555.create_connection()`` which
functions identically to ``socket.create_connection()`` with the same arguments.
This function will automatically fall back on a ``socket.create_connection()`` call if
RFC 6555 is not supported (for instance on platforms not capable of IPv6) or if
RFC 6555 is disabled via setting ``rfc6555.RFC6555_ENABLED`` equal to ``False``.

**IMPORTANT:** Caching is **NOT** thread-safe by default. If you require thread-safe caching
one should create their own implementation of ``rfc6555._RFC6555CacheManager`` object that
is thread-safe and assign an instance to ``rfc6555.cache``.

 .. code-block:: python
 
  import rfc6555
  sock = rfc6555.create_connection(('www.google.com', 80), timeout=10, source_address=('::1', 0))

  # This will disable the Happy Eyeballs algorithm for future
  # calls to create_connection()
  rfc6555.RFC6555_ENABLED = False
  
  # Use this to set a different duration for cache entries.
  rfc6555.cache.validity_duration = 10  # 10 second validity time.

  # Use this to disable caching.
  rfc6555.cache.enabled = False

Support
-------

This module supports Python 2.7 or newer and supports all major platforms.
Additionally if you have ``selectors2>=2.0.0`` installed this module will
also support Jython in addition to CPython.

License
-------

The ``rfc6555`` package is released under the ``Apache-2.0`` license.

See `full license text in LICENSE file <https://github.com/sethmlarson/rfc6555/blob/master/LICENSE>`_ for more information.


CHANGES.rst
Changelog
=========

0.1.0
-----

- Use ``selectors`` instead of ``selectors2`` for Python 3.5+
- Dropped support for Python 2.6, 3.3, and 3.4


tests/test_ipv6.py
import socket

import rfc6555

try:
    from unittest import mock
except (ImportError, AttributeError):
    import mock


def test_ipv6_available():
    assert rfc6555._detect_ipv6()


def test_ipv6_not_available_socket_has_ipv6_false():
    old_has_ipv6 = socket.has_ipv6
    socket.has_ipv6 = False
    assert not rfc6555._detect_ipv6()
    socket.has_ipv6 = old_has_ipv6


def test_ipv6_not_available_socket_exception_on_init():
    with mock.patch("socket.socket") as fake_socket:
        fake_socket.side_effect = OSError

        assert not rfc6555._detect_ipv6()


def test_ipv6_not_available_socket_exception_on_bind():
    sock = mock.Mock()
    with mock.patch("socket.socket") as fake_socket:
        fake_socket.return_value = sock
        sock.bind.side_effect = OSError

        assert not rfc6555._detect_ipv6()


def test_ipv6_not_available_socket_AF_INET6_not_defined():
    old_AF_INET6 = socket.AF_INET6
    try:
        delattr(socket, "AF_INET6")
        assert not rfc6555._detect_ipv6()
    finally:
        socket.AF_INET6 = old_AF_INET6


tests/test_utils.py
import socket

import pytest


def _check_network():
    sock = None
    try:
        sock = socket.create_connection(("www.google.com", 80))
        sock.close()
        return True
    except Exception:
        if sock:
            sock.close()
        return False


requires_network = pytest.mark.skipif(
    not _check_network(), reason="This test requires a network connection."
)


tests/test_create_connection.py
try:
    from unittest import mock
except (ImportError, AttributeError):
    import mock

import socket

import pytest

import rfc6555

from .test_utils import requires_network


class _BasicCreateConnectionTests(object):
    @requires_network
    def test_create_connection_google(self):
        rfc6555.create_connection(("www.google.com", 80))

    @pytest.mark.parametrize("timeout", [None, 5.0])
    def test_create_connection_has_proper_timeout(self, timeout):
        sock = rfc6555.create_connection(("www.google.com", 80), timeout=timeout)

        assert sock.gettimeout() == timeout

    def test_create_connection_with_source_address_calls_bind(self):
        sock = mock.Mock()
        with mock.patch("socket.socket") as fake_socket:
            fake_socket.return_value = sock

            sock.getsockopt.return_value = 0
            sock.connect_ex.return_value = 0
            sock.gettimeout.return_value = None

            try:
                rfc6555.create_connection(("::1", 0), source_address=("::1", 123))
            except Exception:
                pass

            sock.bind.assert_called_with(("::1", 123))

    def test_getaddr_info_empty_list(self):
        with mock.patch("socket.getaddrinfo") as fake_getaddrinfo:
            fake_getaddrinfo.return_value = []

            with pytest.raises(socket.error):
                rfc6555.create_connection(("::1", 0))

    @requires_network
    def test_create_connection_cached_value(self):
        rfc6555.create_connection(("www.google.com", 80))
        rfc6555.create_connection(("www.google.com", 80))


class TestCreateConnectionTestRFC6555Default(_BasicCreateConnectionTests):
    pass


class TestCreateConnectionTestRFC6555Enabled(_BasicCreateConnectionTests):
    def setup_method(self, test_method):
        rfc6555.RFC6555_ENABLED = True


class TestCreateConnectionTestRFC6555Disabled(_BasicCreateConnectionTests):
    def setup_method(self, test_method):
        rfc6555.RFC6555_ENABLED = False


rfc6555.py
# Copyright 2021 Seth Michael Larson
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Python implementation of the Happy Eyeballs Algorithm described in RFC 6555"""

import errno
import socket

try:
    from selectors import EVENT_WRITE, DefaultSelector
except (ImportError, AttributeError):
    from selectors2 import EVENT_WRITE, DefaultSelector

# time.perf_counter() is defined in Python 3.3
try:
    from time import perf_counter
except (ImportError, AttributeError):
    from time import time as perf_counter


# This list is due to socket.error and IOError not being a
# subclass of OSError until later versions of Python.
_SOCKET_ERRORS = (socket.error, OSError, IOError)


# Detects whether an IPv6 socket can be allocated.
def _detect_ipv6():
    if getattr(socket, "has_ipv6", False) and hasattr(socket, "AF_INET6"):
        _sock = None
        try:
            _sock = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
            _sock.bind(("::1", 0))
            return True
        except _SOCKET_ERRORS:
            if _sock:
                _sock.close()
    return False


_HAS_IPV6 = _detect_ipv6()

# These are error numbers for asynchronous operations which can
# be safely ignored by RFC 6555 as being non-errors.
_ASYNC_ERRNOS = set([errno.EINPROGRESS, errno.EAGAIN, errno.EWOULDBLOCK])
if hasattr(errno, "WSAWOULDBLOCK"):
    _ASYNC_ERRNOS.add(errno.WSAWOULDBLOCK)

_DEFAULT_CACHE_DURATION = 60 * 10  # 10 minutes according to the RFC.

# This value that can be used to disable RFC 6555 globally.
RFC6555_ENABLED = _HAS_IPV6

__all__ = ["RFC6555_ENABLED", "create_connection", "cache"]

__version__ = "0.1.0"
__author__ = "Seth Michael Larson"
__email__ = "sethmichaellarson@gmail.com"
__license__ = "Apache-2.0"


class _RFC6555CacheManager(object):
    def __init__(self):
        self.validity_duration = _DEFAULT_CACHE_DURATION
        self.enabled = True
        self.entries = {}

    def add_entry(self, address, family):
        if self.enabled:
            current_time = perf_counter()

            # Don't over-write old entries to reset their expiry.
            if address not in self.entries or self.entries[address][1] > current_time:
                self.entries[address] = (family, current_time + self.validity_duration)

    def get_entry(self, address):
        if not self.enabled or address not in self.entries:
            return None

        family, expiry = self.entries[address]
        if perf_counter() > expiry:
            del self.entries[address]
            return None

        return family


cache = _RFC6555CacheManager()


class _RFC6555ConnectionManager(object):
    def __init__(
        self, address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT, source_address=None
    ):
        self.address = address
        self.timeout = timeout
        self.source_address = source_address

        self._error = None
        self._selector = DefaultSelector()
        self._sockets = []
        self._start_time = None

    def create_connection(self):
        self._start_time = perf_counter()

        host, port = self.address
        addr_info = socket.getaddrinfo(host, port, socket.AF_UNSPEC, socket.SOCK_STREAM)
        ret = self._connect_with_cached_family(addr_info)

        # If it's a list, then these are the remaining values to try.
        if isinstance(ret, list):
            addr_info = ret
        else:
            cache.add_entry(self.address, ret.family)
            return ret

        # If we don't get any results back then just skip to the end.
        if not addr_info:
            raise socket.error("getaddrinfo returns an empty list")

        sock = self._attempt_connect_with_addr_info(addr_info)

        if sock:
            cache.add_entry(self.address, sock.family)
            return sock
        elif self._error:
            raise self._error
        else:
            raise socket.timeout()

    def _attempt_connect_with_addr_info(self, addr_info):
        sock = None
        try:
            for family, socktype, proto, _, sockaddr in addr_info:
                self._create_socket(family, socktype, proto, sockaddr)
                sock = self._wait_for_connection(False)
                if sock:
                    break
            if sock is None:
                sock = self._wait_for_connection(True)
        finally:
            self._remove_all_sockets()
        return sock

    def _connect_with_cached_family(self, addr_info):
        family = cache.get_entry(self.address)
        if family is None:
            return addr_info

        is_family = []
        not_family = []

        for value in addr_info:
            if value[0] == family:
                is_family.append(value)
            else:
                not_family.append(value)

        sock = self._attempt_connect_with_addr_info(is_family)
        if sock is not None:
            return sock

        return not_family

    def _create_socket(self, family, socktype, proto, sockaddr):
        sock = None
        try:
            sock = socket.socket(family, socktype, proto)

            # If we're using the 'default' socket timeout we have
            # to set it to a real value here as this is the earliest
            # opportunity to without pre-allocating a socket just for
            # this purpose.
            if self.timeout is socket._GLOBAL_DEFAULT_TIMEOUT:
                self.timeout = sock.gettimeout()

            if self.source_address:
                sock.bind(self.source_address)

            # Make the socket non-blocking so we can use our selector.
            sock.settimeout(0.0)

            if self._is_acceptable_errno(sock.connect_ex(sockaddr)):
                self._selector.register(sock, EVENT_WRITE)
                self._sockets.append(sock)

        except _SOCKET_ERRORS as e:
            self._error = e
            if sock is not None:
                _RFC6555ConnectionManager._close_socket(sock)

    def _wait_for_connection(self, last_wait):
        self._remove_all_errored_sockets()

        # This is a safe-guard to make sure sock.gettimeout() is called in the
        # case that the default socket timeout is used. If there are no
        # sockets then we may not have called sock.gettimeout() yet.
        if not self._sockets:
            return None

        # If this is the last time we're waiting for connections
        # then we should wait until we should raise a timeout
        # error, otherwise we should only wait >0.2 seconds as
        # recommended by RFC 6555.
        if last_wait:
            if self.timeout is None:
                select_timeout = None
            else:
                select_timeout = self._get_remaining_time()
        else:
            select_timeout = self._get_select_time()

        # Wait for any socket to become writable as a sign of being connected.
        for key, _ in self._selector.select(select_timeout):
            sock = key.fileobj

            if not self._is_socket_errored(sock):

                # Restore the old proper timeout of the socket.
                sock.settimeout(self.timeout)

                # Remove it from this list to exempt the socket from cleanup.
                self._sockets.remove(sock)
                self._selector.unregister(sock)
                return sock

        return None

    def _get_remaining_time(self):
        if self.timeout is None:
            return None
        return max(self.timeout - (perf_counter() - self._start_time), 0.0)

    def _get_select_time(self):
        if self.timeout is None:
            return 0.2
        return min(0.2, self._get_remaining_time())

    def _remove_all_errored_sockets(self):
        socks = []
        for sock in self._sockets:
            if self._is_socket_errored(sock):
                socks.append(sock)
        for sock in socks:
            self._selector.unregister(sock)
            self._sockets.remove(sock)
            _RFC6555ConnectionManager._close_socket(sock)

    @staticmethod
    def _close_socket(sock):
        try:
            sock.close()
        except _SOCKET_ERRORS:
            pass

    def _is_acceptable_errno(self, errno):
        if errno == 0 or errno in _ASYNC_ERRNOS:
            return True
        self._error = socket.error()
        self._error.errno = errno
        return False

    def _is_socket_errored(self, sock):
        errno = sock.getsockopt(socket.SOL_SOCKET, socket.SO_ERROR)
        return not self._is_acceptable_errno(errno)

    def _remove_all_sockets(self):
        for sock in self._sockets:
            self._selector.unregister(sock)
            _RFC6555ConnectionManager._close_socket(sock)
        self._sockets = []


def create_connection(
    address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT, source_address=None
):
    if RFC6555_ENABLED and _HAS_IPV6:
        manager = _RFC6555ConnectionManager(address, timeout, source_address)
        return manager.create_connection()
    else:
        # This code is the same as socket.create_connection() but is
        # here to make sure the same code is used across all Python versions as
        # the source_address parameter was added to socket.create_connection() in 3.2
        # This segment of code is licensed under the Python Software Foundation License
        # See LICENSE: https://github.com/python/cpython/blob/3.6/LICENSE
        host, port = address
        err = None
        for res in socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM):
            af, socktype, proto, _, sa = res
            sock = None
            try:
                sock = socket.socket(af, socktype, proto)
                if timeout is not socket._GLOBAL_DEFAULT_TIMEOUT:
                    sock.settimeout(timeout)
                if source_address:
                    sock.bind(source_address)
                sock.connect(sa)
                return sock

            except socket.error as _:
                err = _
                if sock is not None:
                    sock.close()

        if err is not None:
            raise err
        else:
            raise socket.error("getaddrinfo returns an empty list")
